{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pwd\n",
    "#'/home/vincentwilde/Documents/thesis_1/Testymolo/testymolo/datamolo/scr'\n",
    "source '../../../.venv/bin/activate'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "INSERTS:str = \"./../../testymolo/media/old_sql/VIRUS_1-2004-11-04_Inserts_2.sql\"\n",
    "TABLES: str = \"./../../testymolo/media/old_sql/VIRUS_1-2004-11-04_Tables.sql\"\n",
    "\n",
    "tables_csv: str = \"./../../testymolo/media/tables_csv/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import os\n",
    "\n",
    "\n",
    "## cleanning outfiles\n",
    "for each in os.listdir(tables_csv):\n",
    "    if(each.endswith(\".csv\")):\n",
    "        with open(os.path.join(tables_csv, each), 'w') as handle:\n",
    "            handle.write(\"\")\n",
    "\n",
    "\n",
    "classes:dict = {}\n",
    "\n",
    "with open(INSERTS) as handle:\n",
    "\n",
    "    for line in handle.readlines():\n",
    "\n",
    "        if(line.startswith(\"INSERT\")):\n",
    "            res = re.search(\"INSERT INTO ([^\\s]+) \",line)\n",
    "            tab_name = res.group(1)\n",
    "            if(tab_name in classes):\n",
    "                classes[tab_name] += 1\n",
    "            else:\n",
    "                classes[tab_name] = 1\n",
    "            outfile = os.path.join(tables_csv,tab_name+\".csv\")\n",
    "            res2 = re.search(\"VALUES \\((.*)\\);\",line)\n",
    "            values = r\"{}\".format(res2.group(1))\n",
    "            with open(outfile,'a') as handle:\n",
    "                handle.write(values +'\\n')\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# YATTA !"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### caution when not all rows same nbr of col\n",
    "\n",
    "def custom_csv_parser_to_list(infilepath:str):\n",
    "    with open(infilepath ,'r') as handle:\n",
    "\n",
    "        rows:list = []\n",
    "\n",
    "        for line in handle.readlines():\n",
    "            line = line.strip(';') # removing terminal ';' \n",
    "            values:list = [] # re-initialize value list\n",
    "            val:str = \"\" # current parsing value\n",
    "            inquote:bool = False # re-initialize in quote : False\n",
    "            quote_char:str = '' # ' or \"\n",
    "            for c in line:\n",
    "                if(c == ','): # encounter comma\n",
    "                    if(not inquote): ## and not in quote\n",
    "                        val = val.strip(\"'\")\n",
    "                        val = val.strip('\"')\n",
    "                        if(len(val) > 0):\n",
    "                            values.append(val)\n",
    "                        else:\n",
    "                            values.append(None)\n",
    "                        val = \"\"\n",
    "                        continue\n",
    "                elif(c == '\"' or c == \"'\"): # encounter quote\n",
    "                    if(inquote): # already in quote \n",
    "                        if(quote_char == c):  # encounter ending quote mark\n",
    "                            inquote = False\n",
    "                        else: # encounter the other quote mark \n",
    "                            pass\n",
    "                    else: # encounter starting quote mark\n",
    "                        inquote = True\n",
    "                        quote_char = c\n",
    "                elif((c == ' ' or c == '\\n') and not inquote):  # get rid of whitespaces\n",
    "                    continue\n",
    "                val += c\n",
    "            if(len(val)>0):\n",
    "                val = val.strip(\"'\")\n",
    "                val = val.strip('\"')\n",
    "                values.append(val)\n",
    "\n",
    "            if(len(values) > 0):\n",
    "                rows.append(values)\n",
    "                #print(values)  \n",
    "        \n",
    "        print(\"nbr of col\", set([len(row) for row in rows]))\n",
    "        print(\"nbr of rows\", len(rows))\n",
    "        return rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Biblio_Struct\n",
    "\n",
    "import os\n",
    "\n",
    "Biblio_Struct_h: list = [\"Id_Biblio\", \"PubMed\", \"Title\", \"Auteur\", \"Journal\", \"Date\"]\n",
    "Biblio_Struct_p: str = os.path.join(tables_csv, 'Biblio_Struct.csv')\n",
    "\n",
    "Biblio_Struct: list = custom_csv_parser_to_list(Biblio_Struct_p)\n",
    "print(Biblio_Struct[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### CAZy_DB\n",
    "CAZy_DB_h: list = [\"DB_ac\", \"Protein\", \"DB_nom\", \"Organism\", \"abr\", \"Tax_id\", \"EC\", \"_3D_status\",\n",
    "                   \"Length\", \"Sequence\", \"DB_note\", \"Created\", \"Modified\", \"PP_status\", \"Lib_sort\"]  # 15\n",
    "CAZy_DB_p: str = os.path.join(tables_csv, 'CAZy_DB.csv')\n",
    "\n",
    "CAZy_DB: list = custom_csv_parser_to_list(CAZy_DB_p)\n",
    "print(CAZy_DB[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### CAZy_GB_GP\n",
    "CAZy_GB_GP_h: list = [\"DB_ac\", \"GP_ac\", \"GB_ac\", \"GP_gi\", \"GP_gene\", \"GP_genomic\", \"GP_begin\", \"GP_end\", \"GP_note\"]  # 9\n",
    "CAZy_GB_GP_p: str = os.path.join(tables_csv, 'CAZy_GB_GP.csv')\n",
    "\n",
    "CAZy_GB_GP: list = custom_csv_parser_to_list(CAZy_GB_GP_p)\n",
    "print(CAZy_GB_GP[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### CAZy_PDB\n",
    "CAZy_PDB_h: list = [\"DB_ac\", \"PDB_id\", \"PDB_chain\", \"PDB_begin\", \"PDB_end\", \"PDB_note\", \"PDB_bornModo\"]  # 7\n",
    "CAZy_PDB_p: str = os.path.join(tables_csv, 'CAZy_PDB.csv')\n",
    "\n",
    "CAZy_PDB: list = custom_csv_parser_to_list(CAZy_PDB_p)\n",
    "print(CAZy_PDB[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nbr of col {8}\n",
      "nbr of rows 773\n",
      "['269', '7', None, 'NS4A protein', '1658', '1711', 'EX-HD-F226 ', 'no']\n"
     ]
    }
   ],
   "source": [
    "### CAZy_PP\n",
    "CAZy_PP_h: list = [\"DB_ac\", \"PP_ac\", \"PP_gi\",\n",
    "                   \"PP_gene\", \"PP_begin\", \"PP_end\", \"PP_note\", \"PP_diff\"]  # 8\n",
    "CAZy_PP_p: str = os.path.join(tables_csv, 'CAZy_PP.csv')\n",
    "\n",
    "CAZy_PP: list = custom_csv_parser_to_list(CAZy_PP_p)\n",
    "print(CAZy_PP[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nbr of col {7}\n",
      "nbr of rows 4\n",
      "['68', 'P35944', 'NCAP_PHODV', 'N', '1', '523', 'Reference']\n"
     ]
    }
   ],
   "source": [
    "### CAZy_SP\n",
    "CAZy_SP_h: list = [\"DB_ac\", \"SP_ac\", \"SP_id\",\n",
    "                   \"SP_gene\", \"SP_begin\", \"SP_end\", \"SP_note\"]  # 7\n",
    "CAZy_SP_p: str = os.path.join(tables_csv, 'CAZy_SP.csv')\n",
    "\n",
    "CAZy_SP: list = custom_csv_parser_to_list(CAZy_SP_p)\n",
    "print(CAZy_SP[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## CAZyModO\n",
    "CAZyModO_h: list=[\"URL_family\", \"URL_title\",\n",
    "                 \"URL_body\", \"URL_relative\", \"URL_note\", \"Modified\"] # 6\n",
    "CAZyModO_p: str = os.path.join(tables_csv, 'CAZyModO.csv')\n",
    "CAZyModO: list = custom_csv_parser_to_list(CAZyModO_p)\n",
    "print(CAZyModO[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Cz_EC\n",
    "Cz_EC_h: list=[\"DB_ac\", \"EC\"]  # 2\n",
    "Cz_EC_p: str = os.path.join(tables_csv, 'Cz_EC.csv')\n",
    "Cz_EC: list = custom_csv_parser_to_list(Cz_EC_p)\n",
    "print(Cz_EC[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC_num\n",
    "EC_num_h: list=[\"EC\", \"EC_name\", \"EC_othername\", \"EC_reaction\", \"EC_comment\"]  # 5\n",
    "EC_num_p: str = os.path.join(tables_csv, 'EC_num.csv')\n",
    "EC_num: list = custom_csv_parser_to_list(EC_num_p)\n",
    "print(EC_num[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fam_biblio\n",
    "Fam_biblio_h: list=[\"Family\", \"Id_Biblio\"]  # 2\n",
    "Fam_biblio_p: str = os.path.join(tables_csv, 'Fam_biblio.csv')\n",
    "Fam_biblio: list = custom_csv_parser_to_list(Fam_biblio_p)\n",
    "print(Fam_biblio[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ModO_Composition\n",
    "ModO_Composition_h: list=[\"DB_ac\", \"Mod_num\", \"Family\",\n",
    "                         \"Subf\"]  # 4\n",
    "ModO_Composition_p: str = os.path.join(tables_csv, 'ModO_Composition.csv')\n",
    "ModO_Composition: list = custom_csv_parser_to_list(ModO_Composition_p)\n",
    "print(ModO_Composition[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ModO_CrossRefs\n",
    "ModO_CrossRefs_h: list=[\"Family\", \"Link\", \"URL_base\", \"URL_string\", \n",
    "                       \"Ref_note\"]  # 5\n",
    "ModO_CrossRefs_p: str = os.path.join(tables_csv, 'ModO_CrossRefs.csv')\n",
    "ModO_CrossRefs: list = custom_csv_parser_to_list(ModO_CrossRefs_p)\n",
    "print(ModO_CrossRefs[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ModO_Families\n",
    "ModO_Families_h: list=[\"Family\", \"Family_Name\", \"Family_Activity\", \"Family_Taxo\",\n",
    "                      \"Clan\", \"Family_note\", \"Family_Private_note\", \"Fold\",\n",
    "                      \"ModoS_uniqfct\", \"ModoS_Activity\", \"ModoS_Description\", \"ModoS_Fold\", \"ModoS_note\", \"ModoS_Private_note\",\n",
    "                      \"web_descript\", \"web_status\"]  # 16\n",
    "ModO_Families_p: str = os.path.join(tables_csv, 'ModO_Families.csv')\n",
    "ModO_Families: list = custom_csv_parser_to_list(ModO_Families_p)\n",
    "print(ModO_Families[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ModO_Limits\n",
    "ModO_Limits_h: list = [\"DB_ac\", \"Mod_num\", \"Mod_begin\", \"Mod_end\"]  # 4\n",
    "ModO_Limits_p: str = os.path.join(tables_csv, 'ModO_Limits.csv')\n",
    "ModO_Limits: list = custom_csv_parser_to_list(ModO_Limits_p)\n",
    "print(ModO_Limits[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nbr of col {6}\n",
      "nbr of rows 58\n",
      "['1', 'motif I', 'helicase', 'super family 1', '[FHILVWY]-x-G-x-P-G-x-G-K-[ST]-x(0,2)-[FHILVWY]', 'NULL']\n"
     ]
    }
   ],
   "source": [
    "## Motifs\n",
    "Motifs_h: list = [\"motifs_ac\", \"motifs_name\", \"motifs_activity\", \"motifs_classification\", \"motifs_form\", \"motifs_note\"]\n",
    "Motifs_p: str = os.path.join(tables_csv, 'Motifs.csv')\n",
    "Motifs: list = custom_csv_parser_to_list(Motifs_p)\n",
    "print(Motifs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Organism\n",
    "Organism_h: list = [\"Tax_id\", \"Name\", \"Categorie\",\n",
    "                    \"Classe\", \"Ordre\", \"Fam\", \"SsFam\", \"Genre\", \"Note_org\"]\n",
    "Organism_p: str = os.path.join(tables_csv, 'Organism.csv')\n",
    "Organism: list = custom_csv_parser_to_list(Organism_p)\n",
    "print(Organism[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Prot_biblio\n",
    "\n",
    "Prot_biblio_h: list = [\"DB_ac\", \"Id_Biblio\"]\n",
    "Prot_biblio_p: str = os.path.join(tables_csv, 'Prot_biblio.csv')\n",
    "Prot_biblio: list = custom_csv_parser_to_list(Prot_biblio_p)\n",
    "print(Prot_biblio[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Prot_Infos\n",
    "\n",
    "Prot_Infos_h: list = [\"DB_ac\", \"INFO_indice\", \"INFO_ac\", \"INFO_Nature\", \"INFO\", \"INFO_pmid\"]\n",
    "Prot_Infos_p: str = os.path.join(tables_csv, 'Prot_Infos.csv')\n",
    "Prot_Infos: list = custom_csv_parser_to_list(Prot_Infos_p)\n",
    "print(Prot_Infos[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Prot_MOTIF\n",
    "\n",
    "Prot_MOTIF_h: list = [\"DB_ac\", \"MOTIF_indice\", \"MOTIF_ac\", \"MOTIF_name\", \"MOTIF_begin\", \"MOTIF_end\",\n",
    "                      \"MOTIF\", \"MOTIF_activity\", \"MOTIF_class\", \"MOTIF_note\", \"MOTIF_ref\", \"MOTIF_pmid\"] # 12\n",
    "Prot_MOTIF_p: str = os.path.join(tables_csv, 'Prot_MOTIF.csv')\n",
    "Prot_MOTIF: list = custom_csv_parser_to_list(Prot_MOTIF_p)\n",
    "print(Prot_MOTIF[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Prot_MUT\n",
    "\n",
    "Prot_MUT_h: list = [\"DB_ac\", \"MUT_indice\", \"MUT_ac\", \"MUT\",\n",
    "                      \"MUT_note\", \"MUT_ref\", \"MUT_pmid\"]  # 7\n",
    "Prot_MUT_p: str = os.path.join(tables_csv, 'Prot_MUT.csv')\n",
    "Prot_MUT: list = custom_csv_parser_to_list(Prot_MUT_p)\n",
    "print(Prot_MUT[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Prot_REG\n",
    "\n",
    "Prot_REG_h: list = [\"DB_ac\", \"REG_indice\", \"REG_ac\", \"REG_name\", \"REG_begin\", \"REG_end\",\n",
    "                    \"REG_note\", \"REG_ref\", \"REG_pmid\"]  # 9\n",
    "Prot_REG_p: str = os.path.join(tables_csv, 'Prot_REG.csv')\n",
    "Prot_REG: list = custom_csv_parser_to_list(Prot_REG_p)\n",
    "print(Prot_REG[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Prot_RI\n",
    "\n",
    "Prot_RI_h: list = [\"DB_ac\", \"RI_indice\", \"RI_ac\", \"RI_name\", \"RI_begin\", \"RI_end\",\n",
    "                   \"RI_no_limit\", \"RI_note\", \"RI_ref\", \"RI_pmid\"]  # 10\n",
    "Prot_RI_p: str = os.path.join(tables_csv, 'Prot_RI.csv')\n",
    "Prot_RI: list = custom_csv_parser_to_list(Prot_RI_p)\n",
    "print(Prot_RI[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "Items:dict = {}\n",
    "for item in ModO_Composition:\n",
    "    fam_id = item[2]\n",
    "    if fam_id in Items:\n",
    "        Items[fam_id] += 1\n",
    "    else :\n",
    "        Items[fam_id] = 1\n",
    "# dict[fam.id] = nbr_sequence\n",
    "\n",
    "Counts: dict = {}\n",
    "for k, v in Items.items():\n",
    "    if v in Counts:\n",
    "        Counts[str(v)] += 1\n",
    "    else :\n",
    "        Counts[str(v)] = 1\n",
    "\n",
    "X: list = [int(val) for val in Counts.keys()]\n",
    "Y: list = Counts.values()\n",
    "count: list = list(Items.values())\n",
    "count.sort()\n",
    "\n",
    "print(\"M11 \", Items[\"M11\"])\n",
    "print(\"M12 \", Items[\"M12\"])\n",
    "print(\"M13 \", Items[\"M13\"])\n",
    "\n",
    "#print(X)\n",
    "#print(Y)\n",
    "#plt.scatter(X,Y)\n",
    "\n",
    "print(min(Items.values()),max(Items.values()))\n",
    "print(count[:10], count[-10:])\n",
    "\n",
    "#y_pos = np.arange(len(Items.keys()))\n",
    "#plt.barh(y_pos, Items.values())\n",
    "#plt.yticks(y_pos, Items.keys())\n",
    "#plt.show()\n",
    "\n",
    "\n",
    "#ones:int = 0\n",
    "#for c in count:\n",
    "#    if c == 1:\n",
    "#        ones += 1\n",
    "#print(\"ones : \", ones)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_family(ac:str, _print=True):\n",
    "    result:list = []\n",
    "    for item in ModO_Composition:\n",
    "        if(item[0] == ac):\n",
    "            result.append(item[1:])\n",
    "    if(_print):\n",
    "        print(f\"ac {ac} : {len(result)}\")\n",
    "    return(result)\n",
    "\n",
    "print(get_family(\"191\"))\n",
    "\n",
    "print(get_family(\"3\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nbr_annotations(ac:str, _print=True):\n",
    "    Annotations:list = [Prot_Infos, Prot_MOTIF, Prot_MUT, Prot_REG, Prot_RI, CAZy_GB_GP, CAZy_PDB, CAZy_PP, CAZy_SP]\n",
    "    Annotations_txt:list = [\"Prot_Infos\", \"Prot_MOTIF\", \"Prot_MUT\", \"Prot_REG\", \"Prot_RI\", \"CAZy_GB_GP\", \"CAZy_PDB\", \"CAZy_PP\", \"CAZy_SP\"]\n",
    "    result:list = []\n",
    "    for a, tab in enumerate(Annotations):\n",
    "        for item in tab:\n",
    "            if(item[0] == ac ):\n",
    "                result.append((Annotations_txt[a] , item[1:]))\n",
    "    if(_print) :\n",
    "        print(f\"ac {ac}: {len(result)}\")\n",
    "    return result\n",
    "\n",
    "\n",
    "print(get_nbr_annotations(\"191\"))\n",
    "\n",
    "print(get_nbr_annotations(\"3\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_limits(ac:str, _print=True):\n",
    "    result:list=[]\n",
    "    for item in ModO_Limits:\n",
    "        if(ac == item[0]):\n",
    "            result.append(item[1:])\n",
    "    if(_print) :\n",
    "        print(f\"ac {ac}: {len(result)}\")\n",
    "    return [(int(r[0]), int(r[1]), int(r[2])) for r in result]\n",
    "\n",
    "print(get_limits(\"191\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# nbr of annotations per accession\n",
    "X:list = [item[0] for item in CAZy_DB]\n",
    "Y:list = [len(get_nbr_annotations(x, False)) for x in X]\n",
    "\n",
    "print(min(Y),max(Y))\n",
    "\n",
    "#plt.figure(figsize=(10,6))\n",
    "#plt.scatter(X,Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# nbr of annotation without any annotation\n",
    "\n",
    "voids:list = []\n",
    "for i in range(len(X)):\n",
    "    if(Y[i] == 0):\n",
    "        voids.append(X[i])\n",
    "print(len(voids),'/' , len(X))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def question_1_b():\n",
    "    k = 0\n",
    "    j = 0\n",
    "    for ac in X:\n",
    "        j += 1\n",
    "        Fams = get_family(ac, False)\n",
    "        Lims = get_limits(ac, False)\n",
    "        if(len(Fams) != len(Lims)):\n",
    "            #print(\"ac:\", ac, \" fam:\", len(Fams), \" lim:\", len(Lims))\n",
    "            #print(\"fam:\", [(item[0],item[1]) for item in Fams])\n",
    "            #print(\"lim:\", [(item[0],f\"{item[-2]}:{item[-1]}\") for item in Lims])\n",
    "            k += 1\n",
    "    print(k,\"/\",j)\n",
    "    \n",
    "question_1_b()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# families web status ?\n",
    "\n",
    "# tab: Fam_biblio -> 1 -> ['F35', 1]\n",
    "\n",
    "\n",
    "Fam_web_no_null:list = []\n",
    "for item in ModO_Families:\n",
    "    fam_id:str = item[0]\n",
    "    if(item[-1] not in ['NULL', \"NULL\", '0'] or item[-2] not in ['NULL', \"NULL\", \"nd\", None]):\n",
    "        Fam_web_no_null.append(item)\n",
    "print(len(Fam_web_no_null), len(ModO_Families))\n",
    "\n",
    "print(\"web_descript\",  set([it[-2] for it in Fam_web_no_null]))\n",
    "print(\"web_status\",  set([it[-1] for it in Fam_web_no_null]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# families fold ?\n",
    "\n",
    "\n",
    "Fam_fold: list = []\n",
    "i = ModO_Families_h.index('Fold')\n",
    "print(i)\n",
    "\n",
    "for item in ModO_Families:\n",
    "    #fam_id: str = item[0]\n",
    "    fold_val:str = item[i]\n",
    "    if(fold_val not in ['NULL', \"NULL\", \"0\", None, 'no']):\n",
    "        Fam_fold.append(fold_val)\n",
    "print(len(Fam_fold), len(ModO_Families))\n",
    "\n",
    "print(\"fold\",  set(Fam_fold))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sequence -> famille -> composition & limit \n",
    "#          -> annotation\n",
    "\n",
    "# for each family : \n",
    "#     sequence \n",
    "#         composition & limit\n",
    "\n",
    "# != type of family (family = modulo)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_type(fam_id:str):\n",
    "    result:str = \"\"\n",
    "    for i in range(len(fam_id)):\n",
    "        if not fam_id[i].isdigit():\n",
    "            result += fam_id[i]\n",
    "    return result\n",
    "\n",
    "type_of_family:dict = {}\n",
    "families:list = []\n",
    "Modulo_F:list = []\n",
    "\n",
    "for fam in ModO_Families:\n",
    "    fam_type = get_type(fam[0])\n",
    "    families.append(fam[0])\n",
    "\n",
    "    if(fam_type in type_of_family):\n",
    "        type_of_family[fam_type] += 1\n",
    "    else :\n",
    "        type_of_family[fam_type] = 1\n",
    "\n",
    "    if(fam_type == 'F'):\n",
    "        Modulo_F.append(fam)\n",
    "\n",
    "print(\"len(type_of_family)\" , len(type_of_family) , ' : ',*type_of_family)\n",
    "print(\"len(families) \", len(families))\n",
    "for t, nbr in type_of_family.items():\n",
    "    print(\"type \",t,\" : \" ,nbr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# found nbr sequence of given family \n",
    "# ModO_composition :  0:seq_id  1:limit  2:fam_id  3:subfam\n",
    "\n",
    "def get_fam_composition(fam_id:str=\"\", _print:bool=True):\n",
    "    result:list = []\n",
    "    for item in ModO_Composition:\n",
    "        if(item[2] == fam_id):\n",
    "            result.append(item)\n",
    "    return result\n",
    "\n",
    "# for each in family\n",
    "# print fam_composition\n",
    "\n",
    "get_fam_composition(\"F33\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for seq in get_fam_composition(\"F33\"):\n",
    "    seq_id = seq[0]\n",
    "    print(get_nbr_annotations(seq_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "family  F20  : nbr sequences  35  : annotations  {3, 4, 5, 6, 8}\n",
      "family  F17  : nbr sequences  5  : annotations  {1}\n",
      "family  F11  : nbr sequences  4  : annotations  {1, 5}\n",
      "family  F10  : nbr sequences  2  : annotations  {1}\n",
      "family  F51  : nbr sequences  44  : annotations  {1, 3, 4, 5, 6, 8}\n",
      "family  F52  : nbr sequences  5  : annotations  {3, 4}\n",
      "family  F50  : nbr sequences  23  : annotations  {8, 4, 6}\n",
      "family  F34  : nbr sequences  1  : annotations  {1}\n",
      "family  F21  : nbr sequences  9  : annotations  {4}\n",
      "family  F35  : nbr sequences  4  : annotations  {2, 4}\n",
      "family  F1  : nbr sequences  19  : annotations  {16, 2, 4}\n",
      "family  F2  : nbr sequences  23  : annotations  {16, 2, 4}\n",
      "family  F5  : nbr sequences  5  : annotations  {1}\n",
      "family  F6  : nbr sequences  4  : annotations  {1}\n",
      "family  F8  : nbr sequences  4  : annotations  {1}\n",
      "family  F9  : nbr sequences  3  : annotations  {1}\n",
      "family  F36  : nbr sequences  4  : annotations  {2, 4}\n",
      "family  F12  : nbr sequences  5  : annotations  {4}\n",
      "family  F13  : nbr sequences  3  : annotations  {1, 2}\n",
      "family  F18  : nbr sequences  8  : annotations  {1}\n",
      "family  F16  : nbr sequences  2  : annotations  {1, 2}\n",
      "family  F15  : nbr sequences  4  : annotations  {2}\n",
      "family  F14  : nbr sequences  3  : annotations  {1, 4}\n",
      "family  F23  : nbr sequences  8  : annotations  {1}\n",
      "family  F46  : nbr sequences  2  : annotations  {1, 2}\n",
      "family  F22  : nbr sequences  12  : annotations  {2, 3}\n",
      "family  F24  : nbr sequences  13  : annotations  {1, 2}\n",
      "family  F27  : nbr sequences  5  : annotations  {1, 2, 5}\n",
      "family  F26  : nbr sequences  4  : annotations  {3}\n",
      "family  F45  : nbr sequences  6  : annotations  {1, 2, 5}\n",
      "family  F44  : nbr sequences  1  : annotations  {24}\n",
      "family  F43  : nbr sequences  4  : annotations  {2}\n",
      "family  F25  : nbr sequences  9  : annotations  {1, 2, 3, 5}\n",
      "family  F28  : nbr sequences  4  : annotations  {1, 2, 3}\n",
      "family  F29  : nbr sequences  0  : annotations  set()\n",
      "family  F48  : nbr sequences  3  : annotations  {1}\n",
      "family  F47  : nbr sequences  3  : annotations  {1}\n",
      "family  F66  : nbr sequences  1  : annotations  {3}\n",
      "family  F65  : nbr sequences  1  : annotations  {5}\n",
      "family  F64  : nbr sequences  3  : annotations  {4}\n",
      "family  F49  : nbr sequences  9  : annotations  {8, 4, 6}\n",
      "family  F53  : nbr sequences  2  : annotations  {3, 4}\n",
      "family  F54  : nbr sequences  1  : annotations  {1}\n",
      "family  F55  : nbr sequences  1  : annotations  {2}\n",
      "family  F57  : nbr sequences  1  : annotations  {3}\n",
      "family  F58  : nbr sequences  26  : annotations  {8, 4, 6}\n",
      "family  F59  : nbr sequences  1  : annotations  {5}\n",
      "family  F60  : nbr sequences  1  : annotations  {3}\n",
      "family  F61  : nbr sequences  4  : annotations  {4, 5}\n",
      "family  F62  : nbr sequences  5  : annotations  {3, 4}\n",
      "family  F19  : nbr sequences  4  : annotations  {4, 5}\n",
      "family  F7  : nbr sequences  5  : annotations  {3, 4}\n",
      "family  F3  : nbr sequences  4  : annotations  {4, 5}\n",
      "family  F67  : nbr sequences  2  : annotations  {3, 4}\n",
      "family  F68  : nbr sequences  4  : annotations  {4, 5}\n",
      "family  F69  : nbr sequences  4  : annotations  {4, 5}\n",
      "family  F70  : nbr sequences  3  : annotations  {4}\n",
      "family  F71  : nbr sequences  19  : annotations  {12, 14, 19, 21, 22, 23, 24, 26}\n",
      "family  F72  : nbr sequences  17  : annotations  {14, 19, 21, 22, 23, 24, 26}\n",
      "family  F73  : nbr sequences  6  : annotations  {14, 21, 22, 23, 26}\n",
      "family  F74  : nbr sequences  20  : annotations  {12, 14, 19, 21, 22, 23, 24, 26}\n",
      "family  F75  : nbr sequences  17  : annotations  {14, 19, 21, 22, 23, 24, 26}\n",
      "family  F76  : nbr sequences  17  : annotations  {14, 19, 21, 22, 23, 24, 26}\n",
      "family  F77  : nbr sequences  20  : annotations  {12, 14, 19, 21, 22, 23, 24, 26}\n",
      "family  F78  : nbr sequences  20  : annotations  {12, 14, 19, 21, 22, 23, 24, 26}\n",
      "family  F33  : nbr sequences  3  : annotations  {1}\n",
      "family  F42  : nbr sequences  1  : annotations  {1}\n",
      "family  F56  : nbr sequences  7  : annotations  {1}\n",
      "family  F63  : nbr sequences  1  : annotations  {1}\n",
      "family  F231  : nbr sequences  2  : annotations  {43, 19}\n",
      "family  F232  : nbr sequences  5  : annotations  {1, 2, 5}\n",
      "family  F79  : nbr sequences  20  : annotations  {12, 14, 19, 21, 22, 23, 24, 26}\n",
      "family  F80  : nbr sequences  3  : annotations  {1, 10, 2}\n",
      "family  F81  : nbr sequences  1  : annotations  {2}\n",
      "family  F83  : nbr sequences  1  : annotations  {1}\n",
      "family  F82  : nbr sequences  5  : annotations  {1}\n",
      "family  F84  : nbr sequences  3  : annotations  {1}\n",
      "family  F85  : nbr sequences  1  : annotations  {1}\n",
      "family  F86  : nbr sequences  6  : annotations  {1}\n",
      "family  F87  : nbr sequences  2  : annotations  {1}\n",
      "family  F88  : nbr sequences  2  : annotations  {1}\n",
      "family  F89  : nbr sequences  4  : annotations  {1}\n",
      "family  F90  : nbr sequences  2  : annotations  {1}\n",
      "family  F91  : nbr sequences  2  : annotations  {1}\n",
      "family  F92  : nbr sequences  2  : annotations  {1}\n",
      "family  F93  : nbr sequences  4  : annotations  {1}\n",
      "family  F94  : nbr sequences  1  : annotations  {1}\n",
      "family  F95  : nbr sequences  2  : annotations  {1}\n",
      "family  F96  : nbr sequences  2  : annotations  {1}\n",
      "family  F97  : nbr sequences  1  : annotations  {1}\n",
      "family  F98  : nbr sequences  1  : annotations  {1}\n",
      "family  F99  : nbr sequences  1  : annotations  {1}\n",
      "family  F100  : nbr sequences  1  : annotations  {1}\n",
      "family  F101  : nbr sequences  2  : annotations  {1}\n",
      "family  F102  : nbr sequences  1  : annotations  {1}\n",
      "family  F103  : nbr sequences  2  : annotations  {1}\n",
      "family  F104  : nbr sequences  1  : annotations  {1}\n",
      "family  F105  : nbr sequences  4  : annotations  {1}\n",
      "family  F106  : nbr sequences  4  : annotations  {1}\n",
      "family  F107  : nbr sequences  1  : annotations  {1}\n",
      "family  F108  : nbr sequences  1  : annotations  {4}\n",
      "family  F109  : nbr sequences  1  : annotations  {2}\n",
      "family  F110  : nbr sequences  3  : annotations  {1}\n",
      "family  F111  : nbr sequences  3  : annotations  {1}\n",
      "family  F112  : nbr sequences  6  : annotations  {1, 4}\n",
      "family  F113  : nbr sequences  6  : annotations  {1}\n",
      "family  F114  : nbr sequences  6  : annotations  {1}\n",
      "family  F115  : nbr sequences  6  : annotations  {4}\n",
      "family  F116  : nbr sequences  1  : annotations  {1}\n",
      "family  F117  : nbr sequences  1  : annotations  {1}\n",
      "family  F118  : nbr sequences  3  : annotations  {1, 2}\n",
      "family  F119  : nbr sequences  3  : annotations  {1, 2}\n",
      "family  F120  : nbr sequences  2  : annotations  {1}\n",
      "family  F121  : nbr sequences  2  : annotations  {1}\n",
      "family  F123  : nbr sequences  1  : annotations  {1}\n",
      "family  F124  : nbr sequences  2  : annotations  {1}\n",
      "family  F125  : nbr sequences  1  : annotations  {1}\n",
      "family  F126  : nbr sequences  1  : annotations  {1}\n",
      "family  F127  : nbr sequences  1  : annotations  {1}\n",
      "family  F128  : nbr sequences  1  : annotations  {1}\n",
      "family  F129  : nbr sequences  2  : annotations  {1}\n",
      "family  F130  : nbr sequences  1  : annotations  {1}\n",
      "family  F131  : nbr sequences  1  : annotations  {1}\n",
      "family  F132  : nbr sequences  1  : annotations  {1}\n",
      "family  F133  : nbr sequences  1  : annotations  {1}\n",
      "family  F134  : nbr sequences  1  : annotations  {1}\n",
      "family  F135  : nbr sequences  1  : annotations  {1}\n",
      "family  F136  : nbr sequences  2  : annotations  {1}\n",
      "family  F137  : nbr sequences  1  : annotations  {6}\n",
      "family  F138  : nbr sequences  1  : annotations  {6}\n",
      "family  F139  : nbr sequences  3  : annotations  {6}\n",
      "family  F140  : nbr sequences  3  : annotations  {2, 3}\n",
      "family  F141  : nbr sequences  3  : annotations  {6}\n",
      "family  F142  : nbr sequences  1  : annotations  {6}\n",
      "family  F143  : nbr sequences  3  : annotations  {6}\n",
      "family  F144  : nbr sequences  3  : annotations  {6}\n",
      "family  F145  : nbr sequences  2  : annotations  {6}\n",
      "family  F146  : nbr sequences  1  : annotations  {6}\n",
      "family  F147  : nbr sequences  4  : annotations  {1}\n",
      "family  F148  : nbr sequences  4  : annotations  {1}\n",
      "family  F149  : nbr sequences  3  : annotations  {1}\n",
      "family  F150  : nbr sequences  3  : annotations  {1}\n",
      "family  F151  : nbr sequences  1  : annotations  {1}\n",
      "family  F152  : nbr sequences  1  : annotations  {1}\n",
      "family  F153  : nbr sequences  2  : annotations  {1}\n",
      "family  F154  : nbr sequences  3  : annotations  {1}\n",
      "family  F155  : nbr sequences  1  : annotations  {1}\n",
      "family  F156  : nbr sequences  2  : annotations  {2}\n",
      "family  F157  : nbr sequences  2  : annotations  {1}\n",
      "family  F158  : nbr sequences  1  : annotations  {1}\n",
      "family  F159  : nbr sequences  5  : annotations  {1}\n",
      "family  F160  : nbr sequences  1  : annotations  {2}\n",
      "family  F161  : nbr sequences  2  : annotations  {1}\n",
      "family  F162  : nbr sequences  1  : annotations  {1}\n",
      "family  F163  : nbr sequences  1  : annotations  {1}\n",
      "family  F164  : nbr sequences  1  : annotations  {1}\n",
      "family  F165  : nbr sequences  1  : annotations  {1}\n",
      "family  F166  : nbr sequences  6  : annotations  {20, 21, 22}\n",
      "family  F167  : nbr sequences  1  : annotations  {1}\n",
      "family  F168  : nbr sequences  6  : annotations  {20, 21, 22}\n",
      "family  F169  : nbr sequences  1  : annotations  {1}\n",
      "family  F170  : nbr sequences  1  : annotations  {1}\n",
      "family  F171  : nbr sequences  1  : annotations  {1}\n",
      "family  F172  : nbr sequences  1  : annotations  {1}\n",
      "family  F173  : nbr sequences  1  : annotations  {1}\n",
      "family  F174  : nbr sequences  1  : annotations  {1}\n",
      "family  F175  : nbr sequences  1  : annotations  {7}\n",
      "family  F176  : nbr sequences  1  : annotations  {1}\n",
      "family  F177  : nbr sequences  1  : annotations  {1}\n",
      "family  F178  : nbr sequences  2  : annotations  {1}\n",
      "family  F179  : nbr sequences  3  : annotations  {1}\n",
      "family  F180  : nbr sequences  3  : annotations  {1}\n",
      "family  F181  : nbr sequences  2  : annotations  {1}\n",
      "family  F182  : nbr sequences  2  : annotations  {1, 2}\n",
      "family  F183  : nbr sequences  2  : annotations  {1}\n",
      "family  F184  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F185  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F186  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F187  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F188  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F189  : nbr sequences  12  : annotations  {35, 37, 38, 39, 40, 41, 24, 28, 31}\n",
      "family  F190  : nbr sequences  12  : annotations  {35, 37, 38, 39, 40, 41, 24, 28, 31}\n",
      "family  F191  : nbr sequences  8  : annotations  {37, 38, 39, 40, 41, 31}\n",
      "family  F192  : nbr sequences  12  : annotations  {35, 37, 38, 39, 40, 41, 24, 28, 31}\n",
      "family  F193  : nbr sequences  8  : annotations  {37, 38, 39, 40, 41, 31}\n",
      "family  F194  : nbr sequences  2  : annotations  {10, 37}\n",
      "family  F195  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F196  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F197  : nbr sequences  7  : annotations  {38, 39, 41, 11, 12, 31}\n",
      "family  F198  : nbr sequences  5  : annotations  {37, 38, 40, 10, 12}\n",
      "family  F199  : nbr sequences  2  : annotations  {10, 38}\n",
      "family  F200  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F201  : nbr sequences  5  : annotations  {37, 38, 40, 10, 12}\n",
      "family  F202  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F203  : nbr sequences  14  : annotations  {37, 38, 39, 40, 41, 10, 11, 12, 31}\n",
      "family  F204  : nbr sequences  5  : annotations  {37, 38, 40, 10, 12}\n",
      "family  F205  : nbr sequences  8  : annotations  {1, 2}\n",
      "family  F206  : nbr sequences  8  : annotations  {1, 2}\n",
      "family  F207  : nbr sequences  2  : annotations  {1}\n",
      "family  F208  : nbr sequences  5  : annotations  {37, 38, 40, 10, 12}\n",
      "family  F209  : nbr sequences  2  : annotations  {41, 12}\n",
      "family  F210  : nbr sequences  1  : annotations  {22}\n",
      "family  F211  : nbr sequences  1  : annotations  {22}\n",
      "family  F212  : nbr sequences  1  : annotations  {22}\n",
      "family  F213  : nbr sequences  1  : annotations  {22}\n",
      "family  F214  : nbr sequences  1  : annotations  {22}\n",
      "family  F215  : nbr sequences  2  : annotations  {12, 21}\n",
      "family  F216  : nbr sequences  6  : annotations  {43, 19, 13, 15}\n",
      "family  F217  : nbr sequences  11  : annotations  {14, 19, 21, 22, 24}\n",
      "family  F218  : nbr sequences  2  : annotations  {12, 21}\n",
      "family  F219  : nbr sequences  12  : annotations  {43, 13, 15, 19, 20, 21, 22}\n",
      "family  F220  : nbr sequences  3  : annotations  {24, 35}\n",
      "family  F221  : nbr sequences  4  : annotations  {24, 35, 28}\n",
      "family  F222  : nbr sequences  4  : annotations  {24, 35, 28}\n",
      "family  F223  : nbr sequences  4  : annotations  {24, 35, 28}\n",
      "family  F224  : nbr sequences  4  : annotations  {24, 35, 28}\n",
      "family  F225  : nbr sequences  2  : annotations  {43, 19}\n",
      "family  F226  : nbr sequences  2  : annotations  {43, 19}\n",
      "family  F227  : nbr sequences  2  : annotations  {43, 19}\n",
      "family  F228  : nbr sequences  4  : annotations  {13, 15}\n",
      "family  F229  : nbr sequences  6  : annotations  {43, 19, 13, 15}\n",
      "family  F230  : nbr sequences  6  : annotations  {43, 19, 13, 15}\n",
      "family  F233  : nbr sequences  5  : annotations  {38, 39, 11, 12, 31}\n",
      "family  F234  : nbr sequences  1  : annotations  {35}\n",
      "family  F236  : nbr sequences  1  : annotations  {2}\n",
      "family  F237  : nbr sequences  2  : annotations  {2, 3}\n",
      "family  F238  : nbr sequences  1  : annotations  {2}\n",
      "family  F239  : nbr sequences  4  : annotations  {2, 10}\n",
      "family  F240  : nbr sequences  5  : annotations  {38, 39, 11, 12, 31}\n",
      "family  F241  : nbr sequences  2  : annotations  {14}\n",
      "family  F242  : nbr sequences  3  : annotations  {13}\n",
      "family  F243  : nbr sequences  1  : annotations  {15}\n",
      "family  F244  : nbr sequences  4  : annotations  {24, 35, 28}\n",
      "family  F258  : nbr sequences  1  : annotations  {22}\n",
      "family  F257  : nbr sequences  2  : annotations  {12, 21}\n",
      "family  F256  : nbr sequences  2  : annotations  {12, 21}\n",
      "family  F255  : nbr sequences  19  : annotations  {12, 14, 19, 21, 22, 23, 24, 26}\n",
      "family  F254  : nbr sequences  2  : annotations  {41, 12}\n",
      "family  F253  : nbr sequences  2  : annotations  {41, 12}\n",
      "family  F245  : nbr sequences  2  : annotations  {10, 37}\n",
      "family  F246  : nbr sequences  2  : annotations  {10, 37}\n",
      "family  F247  : nbr sequences  2  : annotations  {10, 37}\n",
      "family  F248  : nbr sequences  5  : annotations  {38, 39, 11, 12, 31}\n",
      "family  F249  : nbr sequences  5  : annotations  {38, 39, 11, 12, 31}\n",
      "family  F250  : nbr sequences  7  : annotations  {38, 39, 41, 11, 12, 31}\n",
      "family  F251  : nbr sequences  3  : annotations  {40, 12, 37}\n",
      "family  F252  : nbr sequences  3  : annotations  {40, 12, 37}\n",
      "family  F122  : nbr sequences  1  : annotations  {5}\n",
      "family  F235  : nbr sequences  6  : annotations  {20, 21, 22}\n",
      "family  F4  : nbr sequences  6  : annotations  {20, 21, 22}\n",
      "family  F40  : nbr sequences  3  : annotations  {1}\n",
      "family  F30  : nbr sequences  6  : annotations  {1}\n",
      "family  F31  : nbr sequences  1  : annotations  {3}\n",
      "family  F32  : nbr sequences  1  : annotations  {3}\n",
      "family  F39  : nbr sequences  2  : annotations  {24, 35}\n",
      "family  F38  : nbr sequences  4  : annotations  {24, 35, 28}\n",
      "family  F37  : nbr sequences  1  : annotations  {28}\n",
      "family  F41  : nbr sequences  2  : annotations  {1}\n",
      "family  F259  : nbr sequences  2  : annotations  {1, 2}\n",
      "family  F260  : nbr sequences  16  : annotations  {9, 20, 14, 15}\n",
      "family  F261  : nbr sequences  16  : annotations  {9, 20, 14, 15}\n",
      "family  F262  : nbr sequences  14  : annotations  {15, 20, 14, 6}\n",
      "family  F263  : nbr sequences  16  : annotations  {9, 20, 14, 15}\n",
      "family  F264  : nbr sequences  1  : annotations  {12}\n",
      "family  F265  : nbr sequences  1  : annotations  {12}\n",
      "family  F266  : nbr sequences  1  : annotations  {12}\n",
      "family  F267  : nbr sequences  1  : annotations  {12}\n",
      "family  F268  : nbr sequences  0  : annotations  set()\n",
      "family  F269  : nbr sequences  16  : annotations  {9, 20, 14, 15}\n",
      "family  F270  : nbr sequences  1  : annotations  {24}\n",
      "family  F271  : nbr sequences  1  : annotations  {24}\n",
      "family  F272  : nbr sequences  1  : annotations  {28}\n",
      "family  F273  : nbr sequences  6  : annotations  {20, 21, 22}\n",
      "family  F274  : nbr sequences  6  : annotations  {20, 21, 22}\n",
      "family  F275  : nbr sequences  6  : annotations  {20, 21, 22}\n",
      "family  F276  : nbr sequences  6  : annotations  {20, 21, 22}\n",
      "family  F277  : nbr sequences  3  : annotations  {9, 10, 17}\n",
      "family  F278  : nbr sequences  3  : annotations  {17, 10, 9}\n",
      "family  F279  : nbr sequences  3  : annotations  {17, 10, 9}\n",
      "family  F280  : nbr sequences  3  : annotations  {17, 10, 9}\n",
      "family  F281  : nbr sequences  3  : annotations  {17, 10, 9}\n",
      "family  F282  : nbr sequences  3  : annotations  {17, 10, 9}\n",
      "family  F284  : nbr sequences  3  : annotations  {17, 10, 9}\n"
     ]
    }
   ],
   "source": [
    "for fam in Modulo_F:\n",
    "    fam_id = fam[0]\n",
    "    _annotations = [len(get_nbr_annotations(seq[0],False)) for seq in get_fam_composition(fam_id,False)]\n",
    "    \n",
    "    print(\"family \", fam_id , \" : nbr sequences \", len(_annotations), \" : annotations \", set(_annotations))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 307)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_limits(famid:str ,seqid:str):\n",
    "    for comp in ModO_Composition:\n",
    "        if(comp[0] == seqid and comp[2] == famid):\n",
    "            for lim in ModO_Limits:\n",
    "                if(lim[0] == seqid and lim[1] == comp[1]):\n",
    "                    return(int(lim[2]), int(lim[3]))\n",
    "    return None\n",
    "\n",
    "get_limits(\"F33\",\"279\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_MOTIF(line:list):\n",
    "    return (int(line[3]), int(line[4]))\n",
    "\n",
    "def annotate_MUT(line:list):\n",
    "    MUTs:str = line[3]\n",
    "    MUTs_list:list = [mut.strip() for mut in MUTs.split(';')]\n",
    "    return MUTs_list\n",
    "\n",
    "def annotate_REG(line:list):\n",
    "    return (int(line[3]), int(line[4]))\n",
    "\n",
    "def annotate_RI(line:list):\n",
    "    return (int(line[3]), int(line[4]))\n",
    "\n",
    "def annotate_GB_GP(line:list):\n",
    "    return line[0]\n",
    "\n",
    "def annotate_PDB(line:list):\n",
    "    return (line[0],line[1])\n",
    "\n",
    "def annotate_PP(line:list):\n",
    "    return int(line[3]), int(line[4])\n",
    "\n",
    "def annotate_SP(line:list):\n",
    "    return int(line[3]), int(line[4])\n",
    "\n",
    "def annotate(line:tuple):\n",
    "    if(line[0] == \"Prot_MOTIF\"):\n",
    "        return annotate_MOTIF(line[1])\n",
    "    elif(line[0] == \"Prot_REG\"):\n",
    "        return annotate_REG(line[1])\n",
    "    elif(line[0] == \"Prot_RI\"):\n",
    "        return annotate_RI(line[1])\n",
    "    elif(line[0] == \"CAZy_PP\"):\n",
    "        return annotate_PP(line[1])\n",
    "    elif(line[0] == \"CAZy_SP\"):\n",
    "        return annotate_SP(line[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def iswithin(lim:tuple, ann_lim:tuple):\n",
    "    #is within  or same ?\n",
    "    if(ann_lim[0] >= lim[0] and ann_lim[1]<=lim[1]):\n",
    "        return True\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2782\n"
     ]
    }
   ],
   "source": [
    "Modulo_list:list = []\n",
    "for fam in ModO_Families:\n",
    "    fam_id = fam[0]\n",
    "    Seqs:list = get_fam_composition(fam_id, False)\n",
    "    for seq in Seqs:\n",
    "        lim = get_limits(fam_id, seq[0])\n",
    "        if(lim != None):\n",
    "            # fam_id, seq_id, seq_begin, seq_end, bool\n",
    "            Modulo_list.append([fam_id, seq[0], lim[0], lim[1], False])\n",
    "        else:\n",
    "            print(\"get_limits(\",fam_id,\", \",seq[0],\")\")\n",
    "\n",
    "\n",
    "def get_index_Modulo_list(famid:str, seqid:str, seq_s:str, seq_e:str):\n",
    "    for i in range(len(Modulo_list)):\n",
    "        item = Modulo_list[i]\n",
    "        if(item[0] == famid and item[1] == seqid and item[2] == str(seq_s) and item[3] == str(seq_e)):\n",
    "            return i\n",
    "    return None\n",
    "\n",
    "\n",
    "print(len(Modulo_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2620\n"
     ]
    }
   ],
   "source": [
    "Seq_annotation_list:list = []\n",
    "for seq in CAZy_DB:\n",
    "    seq_id = seq[0]\n",
    "    annotations:list = get_nbr_annotations(seq_id, False)\n",
    "    for ann in annotations:\n",
    "        # seq_id, annotation_type, annotation_id, bool\n",
    "        Seq_annotation_list.append([seq_id, ann[0], ann[1][0], False])\n",
    "\n",
    "def get_index_seq_annotation_list(seqid:str, anntype:str, annid:str):\n",
    "    for i in range(len(Seq_annotation_list)):\n",
    "        item = Seq_annotation_list[i]\n",
    "        if(item[0] == seqid and item[1] == anntype and item[2] == annid):\n",
    "            return i\n",
    "    return None\n",
    "\n",
    "\n",
    "print(len(Seq_annotation_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Family :  F20\n",
      "nbr seq:  35\n",
      "\t 574\n",
      "\t 6\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (635, 645) ['0', '1', 'motif A', '635', '645', 'A-T-F-x(2)-T-D-L-Q-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (749, 753) ['0', '2', 'motif C', '749', '753', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (715, 727) ['0', '3', 'motif B', '715', '727', 'G-G-I-E-G-L-C-Q-K-L-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_872278.1', 'NC_005036', None, 'L', None, '1', '2204', 'Reference'])\n",
      "\t 23\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (727, 737) ['0', '1', 'motif A', '727', '737', 'S-S-F-x(2)-T-D-L-Q-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (841, 845) ['0', '2', 'motif C', '841', '845', 'Q-G-D-N-E', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (807, 819) ['0', '3', 'motif B', '807', '819', 'G-G-I-E-G-Y-S-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_054697.1', 'NC_002199', '9634976', 'L', None, '1', '2270', 'Reference'])\n",
      "\t 96\n",
      "\t 4\n",
      "\tannotations:  8\n",
      "\t\t  -- Prot_MOTIF (1695, 1712) ['0', '1', 'Hinge', '1695', '1712', 'NULL', 'Hinge', None, 'prouve biochimiquement', 'NULL', '12072530']\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '2', 'motif A', '657', '667', 'S-A-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '3', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '4', 'motif B', '737', '749', 'G-G-I-E-G-Y-C-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('Prot_MUT', ['0', '1', 'mutant abolishing P and L interactions', 'F175A, L176A, F177A, F179A / V201A, L203A, L204A, I205A/L227A, V228A, L229A, M230A/I279A, V280A, M282A, L283A', 'The phosphoprotein (P) and L binding sites reside in the N-terminus of the L subunit of the measles virus RNA polymerase. Virology. 2004 Oct 1;327(2):297-306', '15351217'])\n",
      "\t\t ('Prot_MUT', ['0', '2', 'mutant abolishing P and L interactions refers to 10-20%of WT L', 'I25A, V26A, I28A, L29A', 'The phosphoprotein (P) and L binding sites reside in the N-terminus of the L subunit of the measles virus RNA polymerase. Virology. 2004 Oct 1;327(2):297-306  ', '15351217'])\n",
      "\t\t ('Prot_MUT', ['0', '3', 'mutant abolishing P and L interactions refers to 30-50% WT L', 'E140A, R142A, E143A, D144A/I359A, R360A, F362A, F363A', 'The phosphoprotein (P) and L binding sites reside in the N-terminus of the L subunit of the measles virus RNA polymerase. Virology. 2004 Oct 1;327(2):297-306  ', '15351217'])\n",
      "\t\t ('CAZy_GB_GP', ['NP_056924.1', 'NC_001498', '9626952', 'L', None, '1', '2183', 'Reference'])\n",
      "\t 81\n",
      "\t 8\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '1', 'motif A', '657', '667', 'S-A-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '2', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '3', 'motif B', '737', '749', 'G-G-V-E-G-Y-C-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_047207.1', 'NC_001921', '9630652', None, None, '1', '2184', 'Reference'])\n",
      "\t 89\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (716, 726) ['0', '1', 'motif A', '716', '726', 'S-A-F-x(2)-T-D-L-K-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (830, 834) ['0', '2', 'motif C', '830', '834', 'Q-G-D-N-E', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (796, 808) ['0', '3', 'motif B', '796', '808', 'G-G-I-E-G-Y-S-Q-K-T-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_047113.1', 'NC_001906', '9630565', 'L', None, '1', '2244', 'Reference'])\n",
      "\t 105\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '1', 'motif A', '657', '667', 'S-C-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '2', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '3', 'motif B', '737', '749', 'G-G-I-E-G-Y-C-Q-K-L-W-T-L', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_056879.1', 'NC_001552', '9627228', 'L', None, '1', '2228', 'Reference'])\n",
      "\t 115\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '1', 'motif A', '657', '667', 'S-C-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '2', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '3', 'motif B', '737', '749', 'G-G-I-E-G-Y-C-Q-K-L-W-T-L', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_604442.1', 'NC_003461', '19718373', 'L', None, '1', '2223', 'Reference'])\n",
      "\t 209\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (613, 623) ['0', '1', 'motif A', '613', '623', 'T-Y-A-x(2)-L-D-Y-E-K-W', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (728, 732) ['0', '2', 'motif C', '728', '732', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (694, 706) ['0', '3', 'motif B', '694', '706', 'G-G-L-E-G-L-R-Q-K-G-W-S-L', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_478343.1', 'NC_003243', '17158073', 'L', None, '1', '2128', 'Reference'])\n",
      "\t 123\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (682, 692) ['0', '1', 'motif A', '682', '692', 'S-C-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (796, 800) ['0', '2', 'motif C', '796', '800', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (762, 774) ['0', '3', 'motif B', '762', '774', 'G-G-I-E-G-F-C-Q-K-L-W-T-L', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_067153.1', 'NC_001796', '10937876', None, None, '1', '2258', 'Reference'])\n",
      "\t 37\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '1', 'motif A', '657', '667', 'A-S-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '2', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '3', 'motif B', '737', '749', 'G-G-I-E-G-L-C-Q-K-A-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['AAC95518.1', 'AF052755', '2981093', 'L', None, '1', '2255', 'Reference'])\n",
      "\t 15\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (662, 672) ['0', '1', 'motif A', '662', '672', 'A-S-F-x(2)-T-D-L-Q-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (776, 780) ['0', '2', 'motif C', '776', '780', 'Q-G-E-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (742, 754) ['0', '3', 'motif B', '742', '754', 'G-G-I-E-G-L-C-Q-K-M-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_150063.1', 'NC_003043', '15081574', 'L', None, '1', '2241', 'Reference'])\n",
      "\t 29\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (667, 677) ['0', '1', 'motif A', '667', '677', 'A-C-F-x(2)-T-D-L-T-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (781, 785) ['0', '2', 'motif C', '781', '785', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (747, 759) ['0', '3', 'motif B', '747', '759', 'G-G-I-E-G-L-C-Q-K-M-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['CAA45569.1', 'X64275', '433522', 'L', None, '1', '2269', 'Reference'])\n",
      "\t 46\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (663, 673) ['0', '1', 'motif A', '663', '673', 'A-C-F-x(2)-T-D-L-T-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (777, 781) ['0', '2', 'motif C', '777', '781', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (743, 755) ['0', '3', 'motif B', '743', '755', 'G-G-I-E-G-L-C-Q-K-L-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['AAL83746.1', 'AF345290', '19070177', 'L', None, '1', '2261', 'Reference'])\n",
      "\t 54\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (663, 673) ['0', '1', 'motif A', '663', '673', 'A-C-F-x(2)-T-D-L-T-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (777, 781) ['0', '2', 'motif C', '777', '781', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (743, 755) ['0', '3', 'motif B', '743', '755', 'G-G-I-E-G-L-C-Q-K-L-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_054714.1', 'NC_002200', '9695423', 'L', None, '1', '2261', 'Reference'])\n",
      "\t 67\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (635, 645) ['0', '1', 'motif A', '635', '645', 'A-T-F-x(2)-T-D-L-Q-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (749, 753) ['0', '2', 'motif C', '749', '753', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (715, 727) ['0', '3', 'motif B', '715', '727', 'G-G-I-E-G-L-C-Q-K-L-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_071471.1', 'NC_002617', '11545725', 'L', None, '1', '2204', 'Reference'])\n",
      "\t 129\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '1', 'motif A', '657', '667', 'S-C-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '2', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '3', 'motif B', '737', '749', 'G-G-I-E-G-F-C-Q-K-L-W-T-L', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_037646.1', 'NC_002161', '9634115', None, None, '1', '2233', 'Reference'])\n",
      "\t 161\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (692, 701) ['0', '1', 'motif A', '692', '701', 'C-S-I-x-T-D-L-S-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (806, 810) ['0', '2', 'motif C', '806', '810', 'N-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (772, 784) ['0', '3', 'motif B', '772', '784', 'G-G-I-E-G-W-C-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_048058.1', 'NC_001989', '9631278', 'L', None, '1', '2162', 'Reference'])\n",
      "\t 180\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (627, 636) ['0', '1', 'motif A', '627', '636', 'S-S-F-x-T-D-L-E-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (740, 744) ['0', '2', 'motif C', '740', '744', 'M-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (706, 718) ['0', '3', 'motif B', '706', '718', 'G-G-I-E-G-L-Q-Q-K-L-W-T-S', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_066251.1', 'NC_002549', '10313999', 'L', None, '1', '2212', 'Reference'])\n",
      "\t 194\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (599, 609) ['0', '1', 'motif A', '599', '609', 'C-I-A-x(2)-I-D-Y-E-K-W', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (712, 716) ['0', '2', 'motif C', '712', '716', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (678, 690) ['0', '3', 'motif B', '678', '690', 'G-G-L-E-G-L-R-Q-K-G-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_041716.1', 'NC_001560', '9627234', None, None, '1', '2109', 'Reference'])\n",
      "\t 204\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (612, 622) ['0', '1', 'motif A', '612', '622', 'T-Y-A-x(2)-L-D-Y-E-K-W', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (727, 731) ['0', '2', 'motif C', '727', '731', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (693, 705) ['0', '3', 'motif B', '693', '705', 'G-G-L-E-G-L-R-Q-K-G-W-S-L', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_056797.1', 'NC_001542', '9627202', None, None, '1', '2142', 'Reference'])\n",
      "\t 221\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (625, 635) ['0', '1', 'motif A', '625', '635', 'T-I-A-x(2)-I-D-Y-E-K-W', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (738, 742) ['0', '2', 'motif C', '738', '742', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (704, 716) ['0', '3', 'motif B', '704', '716', 'G-G-L-E-G-L-R-Q-K-G-W-S-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_065409.1', 'NC_002526', '10086573', 'L', None, '1', '2144', 'Reference'])\n",
      "\t 199\n",
      "\t 4\n",
      "\tannotations:  3\n",
      "\t\t  -- Prot_MOTIF (588, 598) ['0', '1', 'motif A', '588', '598', 'C-I-A-x(2)-I-D-Y-E-K-W', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (701, 705) ['0', '2', 'motif C', '701', '705', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_116748.1', 'NC_002803', '14336459', 'L', None, '1', '2095', 'Reference'])\n",
      "\t 171\n",
      "\t 3\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (695, 704) ['0', '1', 'motif A', '695', '704', 'C-S-I-x-T-D-L-S-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (809, 813) ['0', '2', 'motif C', '809', '813', 'N-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (775, 787) ['0', '3', 'motif B', '775', '787', 'G-G-I-E-G-W-C-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_044598.1', 'NC_001803', '9629377', 'L', None, '1', '2165', 'Reference'])\n",
      "\t 150\n",
      "\t 4\n",
      "\tannotations:  5\n",
      "\t\t  -- Prot_MOTIF (695, 704) ['0', '1', 'motif A', '695', '704', 'C-S-I-x-T-D-L-S-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (809, 813) ['0', '2', 'motif C', '809', '813', 'N-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (775, 787) ['0', '3', 'motif B', '775', '787', 'G-G-I-E-G-W-C-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('Prot_MUT', ['0', '1', 'N1049x', 'A single amino acid mutation in position 1049N specifically affects the ability of the polymeraseto terminate transcription, but which does not affect genome replication or interaction with the M2-1 protein.', 'J Virol. 2003 Jul;77(13):7352-60', '12805433'])\n",
      "\t\t ('CAZy_GB_GP', ['NP_056866.1', 'NC_001781', '9629209', 'L', None, '1', '2166', 'Reference'])\n",
      "\t 187\n",
      "\t 5\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (630, 639) ['0', '1', 'motif A', '630', '639', 'A-S-F-x-T-D-L-E-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (743, 747) ['0', '2', 'motif C', '743', '747', 'M-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (709, 721) ['0', '3', 'motif B', '709', '721', 'G-G-I-E-G-L-Q-Q-K-L-W-T-C', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_042031.1', 'NC_001608', '13489282', 'L', None, '1', '2331', 'Reference'])\n",
      "\t 366\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '1', 'motif A', '657', '667', 'S-A-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '2', 'motif B', '737', '749', 'G-G-I-E-G-Y-C-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '3', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_945030.1', 'NC_005283', None, 'L', None, '1', '2183', 'Reference'])\n",
      "\t 374\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (661, 671) ['0', '1', 'motif A', '661', '671', 'S-T-F-x(2)-T-D-L-Q-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (775, 779) ['0', '2', 'motif C', '775', '779', 'Q-G-D-N-E', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (741, 753) ['0', '3', 'motif B', '741', '753', 'G-G-I-E-G-F-N-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', 'WARNING : G-G-I-E-G-F-[N]instead_of[CSQR]-Q-K-L-W-T-I', None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_958055.1', 'NC_005339', None, 'L', None, '1', '2205', 'Reference'])\n",
      "\t 353\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '1', 'motif A', '657', '667', 'S-A-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '2', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '3', 'motif B', '737', '749', 'G-G-I-E-G-Y-C-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['AAF85698.1', 'AF266290', None, 'L', None, '1', '2183', 'Reference'])\n",
      "\t 401\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (627, 636) ['0', '1', 'motif A', '627', '636', 'S-S-F-x-T-D-L-E-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (740, 744) ['0', '2', 'motif C', '740', '744', 'M-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (706, 718) ['0', '3', 'motif B', '706', '718', 'G-G-I-E-G-L-Q-Q-K-L-W-T-S', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_690587.1', 'NC_004161', None, 'L', None, '1', '2212', 'Reference'])\n",
      "\t 414\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (630, 639) ['0', '1', 'motif A', '630', '639', 'A-S-I-x-T-D-L-S-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (709, 721) ['0', '2', 'motif B', '709', '721', 'G-G-I-E-G-W-C-Q-K-L-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (743, 747) ['0', '3', 'motif C', '743', '747', 'N-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_690071.1', 'NC_004148', None, 'L', None, '1', '2005', 'Reference'])\n",
      "\t 61\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (662, 672) ['0', '1', 'motif A', '662', '672', 'A-C-F-x(2)-T-D-L-A-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (775, 779) ['0', '2', 'motif C', '775', '779', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (741, 753) ['0', '3', 'motif B', '741', '753', 'G-G-I-E-G-L-C-Q-K-M-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_598406.1', 'NC_003443', '19525727', 'L', None, '1', '2262', 'Reference'])\n",
      "\t 8\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (716, 726) ['0', '1', 'motif A', '716', '726', 'S-A-F-x(2)-T-D-L-K-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (830, 834) ['0', '2', 'motif C', '830', '834', 'Q-G-D-N-E', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (796, 808) ['0', '3', 'motif B', '796', '808', 'G-G-I-E-G-Y-S-Q-K-T-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_112028.1', 'NC_002728', '13559815', 'L', None, '1', '2244', 'Reference'])\n",
      "\t 344\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (669, 679) ['0', '1', 'motif A', '669', '679', 'A-C-F-x(2)-T-D-L-S-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (783, 787) ['0', '2', 'motif C', '783', '787', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (749, 761) ['0', '3', 'motif B', '749', '761', 'G-G-I-E-G-L-C-Q-K-L-W-T-M', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_665871.1', 'NC_004074', None, 'L', None, '1', '2271', 'Reference'])\n",
      "\t 619\n",
      "\t 4\n",
      "\tannotations:  4\n",
      "\t\t  -- Prot_MOTIF (652, 662) ['0', '1', 'motif A', '652', '662', 'A-S-F-x(2)-T-D-L-Q-K-F', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (766, 770) ['0', '2', 'motif C', '766', '770', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (732, 744) ['0', '3', 'motif B', '732', '744', 'G-G-I-E-G-Y-S-Q-K-M-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['NP_899661.1', 'NC_005084', None, 'L', None, '1', '2181', 'Reference'])\n",
      "\t 74\n",
      "\t 4\n",
      "\tannotations:  6\n",
      "\t\t ('Prot_Infos', ['0', '1', 'Origine sequence', 'ncbi donne un fragment, seq de SP rentree;', 'NULL'])\n",
      "\t\t  -- Prot_MOTIF (657, 667) ['0', '1', 'motif A', '657', '667', 'S-A-F-x(2)-T-D-L-K-K-Y', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (737, 749) ['0', '2', 'motif B', '737', '749', 'G-G-V-E-G-Y-C-Q-K-L-W-T-I', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t  -- Prot_MOTIF (771, 775) ['0', '3', 'motif C', '771', '775', 'Q-G-D-N-Q', 'polymerase', 'mononegavirales', None, None, 'NULL']\n",
      "\t\t ('CAZy_GB_GP', ['BAA01208.1', 'D10371', '222253', 'L', None, '1', '10', ''])\n",
      "\t\t  -- CAZy_SP (1, 2184) ['P90282', 'P90282', 'LARG', '1', '2184', 'Reference']\n"
     ]
    }
   ],
   "source": [
    "# for each family of type 'F'\n",
    "for fam in Modulo_F:\n",
    "    fam_id = fam[0]\n",
    "    print(\"Family : \", fam_id)\n",
    "    Seqs:list = get_fam_composition(fam_id, False)\n",
    "    print(\"nbr seq: \", len(Seqs))\n",
    "    \n",
    "    # for each sequences in each family\n",
    "    for seq in Seqs:\n",
    "        seq_id = seq[0]\n",
    "\n",
    "        # get limits\n",
    "        seq_limits = get_limits(fam_id, seq_id)\n",
    "\n",
    "        # is already delt with ? \n",
    "        k:int = get_index_Modulo_list(fam_id, seq_id, seq_limits[0], seq_limits[1])\n",
    "        if(k != None):\n",
    "            # already done\n",
    "            if(Modulo_list[k][-1] != False):\n",
    "                continue\n",
    "        # else:\n",
    "        # go on then !\n",
    "        print(\"\\t\",seq_id)\n",
    "        \n",
    "        # get annotations\n",
    "        _annotations:list = get_nbr_annotations(seq_id, False)\n",
    "        print(\"\\tannotations: \", len(_annotations))\n",
    "\n",
    "        # for each annotation\n",
    "        for ann in _annotations:\n",
    "\n",
    "            # automatized annotation types\n",
    "            if(ann[0] in [\"Prot_MOTIF\", \"Prot_REG\", \"Prot_RI\", \"CAZy_PP\", \"CAZy_SP\"]):\n",
    "\n",
    "                # if same or with in limits\n",
    "                if(iswithin(seq_limits, annotate(ann))):\n",
    "                    # add/associate annotation to families\n",
    "                    print(\"\\t\\t\",\"<--\",ann[0], \"ok\")\n",
    "                    i:int = get_index_seq_annotation_list()\n",
    "                    if(i != None):\n",
    "                        if(Seq_annotation_list[i][-1] != False):\n",
    "                            # could be multiple families \n",
    "                            Seq_annotation_list[i][-1] += (\" \"+fam_id)\n",
    "                        else:\n",
    "                            Seq_annotation_list[i][-1] = fam_id\n",
    "                    else:\n",
    "                        # should not be\n",
    "                        print(\"annotation: \", seq_id, ann[0], ann[1][0])\n",
    "                else:\n",
    "                    # not same nor within limits\n",
    "                    print(\"\\t\\t\",\" --\",ann[0], annotate(ann) ,ann[1])\n",
    "            else:\n",
    "                # see those annotations later\n",
    "                print(\"\\t\\t\",ann)\n",
    "        #\n",
    "        # break # for each seq\n",
    "    #\n",
    "    #\n",
    "    \n",
    "    \n",
    "    break # for each fam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "with open('Modulo_list', 'w') as file:\n",
    "    handle =  csv.writer(file)\n",
    "    handle.writerows(Modulo_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "with open('Seq_annotation_list', 'w') as file:\n",
    "    handle =  csv.writer(file)\n",
    "    handle.writerows(Seq_annotation_list)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
